sylt-compiler/src/name_resolution.rs

```diff
@@ -0,0 +1,1038 @@
+        // TODO(ed): Find the closest matching name if we don't find anything?
+        for (var_name, var_id) in self.stack.iter().rev() {
+            if var_name == name {
+                return Ok(*var_id);
+            }
+        }
+        match self.lookup_global(namespace_id, name) {
+            Some(Name::Name(var)) => Ok(*var),
+            Some(Name::Namespace(..)) => Err(vec![resolution_error!(
+                self,
+                span,
+                "When resolving the name {:?} - a namespace was found",
+                name
+            )]),
+            None => Err(vec![resolution_error!(
+                self,
+                span,
+                "Failed to resolve {:?} - nothing matched",
+                name
+            )]),
+        }
+    }
+
+    fn namespace_list(&self, namespace_id: usize, assignable: &ParserAssignable) -> Option<usize> {
+        use sylt_parser::AssignableKind as AK;
+        (match &assignable.kind {
+            AK::Read(ident) => match self.lookup_global(namespace_id, &ident.name) {
+                Some(Name::Namespace(file_or_lib, _)) => Some(file_or_lib),
+                _ => None,
+            },
+            AK::Access(prev, ident) => {
+                match self
+                    .namespace_list(namespace_id, prev)
+                    .map(|new_namespace| self.lookup_global(new_namespace, &ident.name))
+                    .flatten()
+                {
+                    Some(Name::Namespace(file_or_lib, _)) => Some(file_or_lib),
+                    _ => None,
+                }
+            }
+            AK::Index { .. }
+            | AK::Variant { .. }
+            | AK::Call { .. }
+            | AK::ArrowCall { .. }
+            | AK::Expression { .. } => None,
+        })
+        .map(|file_or_lib| self.file_to_namespace.get(file_or_lib).cloned())
+        .flatten()
+    }
+
+    fn type_vec(&self, parser_tys: &[ParserType]) -> ResolveResult<Vec<Type>> {
+        let mut tys = Vec::new();
+        let mut errs = Vec::new();
+        for ty in parser_tys.iter() {
+            match self.ty(ty) {
+                Err(mut es) => errs.append(&mut es),
+                Ok(ty) => tys.push(ty),
+            }
+        }
+        if errs.is_empty() {
+            Ok(tys)
+        } else {
+            Err(errs)
+        }
+    }
+
+    fn ty(&self, ty: &ParserType) -> ResolveResult<Type> {
+        use sylt_parser::TypeKind as TK;
+        use Type as T;
+        let span = ty.span;
+        Ok(match &ty.kind {
+            TK::Implied => T::Implied(span),
+            TK::Resolved(t) => T::Resolved(t.clone(), span),
+            TK::UserDefined(t, gs) => {
+                let r = match self.ty_assignable(t)? {
+                    T::UserType(r, _) => r,
+                    _ => raise_resolution_error! {
+                        self,
+                        span,
+                        "This is not a reference to a user defined type"
```

Comment:
I think this is too much information, it would probably suffice to say "This is not a type". They might have misspelled a built-in type, so the desired type needn't be user defined.
```sylt
let x: boolean : true
```

Reply:
Upon further reading, this looks like dead code? `ty_assignable` only ever returns `UserType` as far as I can tell.

Reply:
I was kinda safeguarding against future changes. Right now an `unreachable` would be correct, but it might backfire later. I'm maybe being too defensive though.

But I explicitly wanted it to say that it wasn't a user defined type. So I'm still leaning towards keeping the error message. Since it would still be an error if the type was written as `int`. But I might be rambling even more right now.